[wasm] Implement partial backward branch support in the Jiterpreter

kg opened the pull request:
This PR adds partial support for backward branches to the jiterpreter.

Due to WASM's quirky approach to constrained control flow and loops, the code it generates is very suboptimal, but it still produces a measurable speed-up (~22sec/iter -> ~20sec/iter for one of the benchmarks that regressed, for example) and is a decent starting point to improve on.

This PR also removes a write barrier from ldelem_ref that didn't need to be there and was *very* expensive. That takes the regressed benchmark down further from ~20sec/iter to ~3sec/iter.

Additional statistics and a runtime option are added to go with the new backward branch support.

More detail on how it works:
1. For any methods that contain a trace entry point, the interpreter maintains a table listing all the bblocks that are targeted by a backwards branch.
2. If a method has one or more backward branch targets in its table, when the jiterpreter generates traces in that method it wraps them in a wasm `loop` instruction so that we can transfer control back to the top.
3. The jiterpreter scans the backward branch table when emitting instructions, and when it hits something it knows is a backwards branch target, it ensures that a branch target block starts there. Each encountered backwards branch target is added to a list.
4. Any time the jiterpreter encounters a backwards branch, if the target is in the list of branch targets we've already encountered, `eip` is updated and control is sent back to the top of the trace by branching to the `loop`. After that, execution will skip over blocks until it reaches the branch target.

Attentive readers will note that this algorithm is inefficient because we have to scan over blocks until we find the branch target - there's only one loop for the entire trace. I tried creating a separate loop for each backwards branch target (which would allow us to jump directly to it) but its interaction with forward branches (which require the ability to jump forward to the end of a control region) made it too hard to get that working, at least initially.

A better implementation of control flow would probably allow direct branching for both forwards and backwards control flow, or at least reduce the cost of the branches from what it is currently. But that implementation would likely require a second pass in the trace compiler and building some sort of CFG on the fly so I haven't started on it yet.

kg committed:
Checkpoint backward branch support

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Checkpoint

kg committed:
Turn early trace abort off when back branches are on since it impairs looping
Remove unnecessary slow write barrier from ldelem_ref
Update heuristic for backward branches
Add back branch success rate statistic

vargaz commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -8467,6 +8467,24 @@ generate_compacted_code (TransformData *td)
+				// Most methods in System.Runtime.Tests have 8 or fewer branch targets,
+				//  and many have more than 4, so this is a decent starting size
+				rtm->backward_branch_offsets_size = 8;
+				rtm->backward_branch_offsets = g_malloc0(sizeof(guint16) * rtm->backward_branch_offsets_size);
```
Would be better to keep this in a local variable, and allocate it using mono_mem_manager_alloc0 () at the end of this function, most of the memory belonging to InterpMethod's is allocated that way.

kg replied:
so like stackalloc a big enough scratch buffer, then copy it to a perfectly sized one allocated with the mem manager? that makes sense

vargaz replied:
Its ok to malloc it and free it after copying it to the final memory as well.

kg replied:
isn't that mempool freed at the end of compilation? the jiterp needs it after compilation

vargaz replied:
td->mem_manager is not freed.

kg committed:
Improve back branch offset table allocation
Don't generate a loop for trace if all the back branch offsets are before its start offset
